### ISSUE TITLE
In rails 5.1, the words "before" and "after" are logged as if they are callbacks

### ISSUE THREAD

**Issue by jaredbeck** [jaredbeck/rails-callback_log#4]
In rails 5.1, callback logging is working, but the words "before" and "after" are mixed in with legitimate callbacks.

```
Callback: serialize_data!
Callback: raise_on_session_data_overflow!
Callback: call
Callback: instance_exec
Callback: before
Callback: disable_query_cache!
Callback: after
Callback: before
Callback: before
Callback: before
```

I think this has something to do with the new `CallTemplate` class being more generic than I thought.

**Comment by DexterTheDragon** [comment]
I was curious what was happening with this so I did some digging and here are my findings. First let's compare the log from Rails 5.0 to Rails 5.1.

![compare](https://user-images.githubusercontent.com/96819/46180471-d5e42780-c27d-11e8-99c2-4cba72c4f623.png)
They are practically identical but on Rails 5.0 it prints an object identifier and on Rails 5.1 it's instead printing either `before`, `after`, or `call`. 

If we look at the [rails-callback_log code for Rails 5.0](https://github.com/jaredbeck/rails-callback_log/blob/b6bc97a50edc7d3e9cd129dd7f2ebf5c71111be1/lib/rails-callback_log.rb#L45-L54), we see that it wraps the `make_lambda` method and returns a lambda that logs the callback then runs the original lambda. The message of the callback log uses the `filter` parameter passed to the `make_lambda` method.

If we look at the the Rails 5.0 code, we see that Rails calls [an `apply` method](https://github.com/rails/rails/blob/5a0fce7d4ecd152e883877bc5a6a11af54207fde/activesupport/lib/active_support/callbacks.rb#L350-L362) which calls the `make_lambda` method passing it the `@filter` parameter to create a variable called `user_callback`. That lambda then gets pushed onto a stack to be executed later. Looking at the [code for `make_lambda`](https://github.com/rails/rails/blob/5a0fce7d4ecd152e883877bc5a6a11af54207fde/activesupport/lib/active_support/callbacks.rb#L370-L408) we see that the `filter` parameter can be a symbol, string, proc, or object. This is then passed through a case statement and a lambda created based on it's type.

Now If we look at the [rails-callback_log code for Rails 5.1](https://github.com/jaredbeck/rails-callback_log/blob/b6bc97a50edc7d3e9cd129dd7f2ebf5c71111be1/lib/rails-callback_log.rb#L30-L39), it again wraps the `make_lambda` method but is instead logging the `@method_name` variable.

Jumping to the Rails 5.1 code we again [start with the `apply` method](https://github.com/rails/rails/blob/bb22fe9d4a6102d2a28cb1adfd6fe9d38fc9bb22/activesupport/lib/active_support/callbacks.rb#L333-L345), this time the `user_callback` variable is created by calling `CallTemplate.build(@filter, self)`. Looking at the [code for CallTemplate.build](https://github.com/rails/rails/blob/bb22fe9d4a6102d2a28cb1adfd6fe9d38fc9bb22/activesupport/lib/active_support/callbacks.rb#L435-L456) we see the familiar case statement based on the filter parameter but this time it's instantiating CallTemplate instead of creating a lambda. Looking at the [`initialize` method for CallTemplate](https://github.com/rails/rails/blob/bb22fe9d4a6102d2a28cb1adfd6fe9d38fc9bb22/activesupport/lib/active_support/callbacks.rb#L370-L375) we that this is where `@method_name` gets set. Jumping back to the [apply method](https://github.com/rails/rails/blob/bb22fe9d4a6102d2a28cb1adfd6fe9d38fc9bb22/activesupport/lib/active_support/callbacks.rb#L333-L345) we see that `user_callback.make_lambda` is called when pushing the lambda onto the stack.

The summary is that on Rails 5.0 `apply` calls `make_lambda` with a filter object and rails-callback_log inserts itself here logging the `filter` parameter when the lambda gets executed. On Rails 5.1 it runs `apply` that calls `CallTemplate.build` with the filter object which gets converted into parameters for CallTemplate.new. The `CallTemplate#make_lambda` method is eventually called and rails-callback_log inserts itself but only has the `@method_name` variable available for logging once the lambda is executed. 

You can see this in the above image, when the filter option is a symbol, the Rails 5.0 code directly logs the `filter` variable which prints the symbol as a string ala `configure_query_cache!`. Under 5.1 that symbol gets passed through `CallTemplate.build` which passes it as the `method_name` parameter to `CallTemplate.new` and it gets logged similarly. But when filter is an object the 5.0 code logs the object which prints as an object identifier. With 5.1 `CallTemplate.build` destructs `filter` passing the `filter` as `target` and `method_to_call` as `method_name`. This is why these lines end up getting logged as `before`, `after`, or `call`. 

I'm a little unsure what the best fix would be. A quickfix that seems to work is to change 
```ruby
::RailsCallbackLog.logger.debug(format("Callback: %s", @method_name))
``` 
to 
```ruby
::RailsCallbackLog.logger.debug(format("Callback: %s", @override_target || @method_name))
```
`@override_target` is nil when `filter` is a symbol, string, or proc and it gets set to the value of `filter` when `filter` is an object. The works well for everything except proc which passes things in a completely different manner. 

This examination also skips some points such as the `apply` method on 5.0 using `make_lambda` for the :before, :after, and :around filters, but 5.1 using `make_lambda` for :before and :after and a different setup for :around. 

I think the best fix may lie in digging into [`CallTemplate#expand`](https://github.com/rails/rails/blob/bb22fe9d4a6102d2a28cb1adfd6fe9d38fc9bb22/activesupport/lib/active_support/callbacks.rb#L377-L406) as a better hook point for rails-callback_log to use.

Hope this helps.

**Comment by jaredbeck** [comment]
Thanks for that analysis Kevin! Please go ahead and make a PR.

**Comment by jaredbeck** [comment]
Fixed by https://github.com/jaredbeck/rails-callback_log/pull/5, released in 0.3.0
